### PR TITLE
MVP EMPI implementation 

### PR DESCRIPTION
**Pull request by tadgh**
This is the MVP implementation of EMPI for HAPI, providing the ability for rule-based matching between patients/practitioners to related Persons. 

**30000 foot overview:** 

* Details concerning the implementation can be found here
* Add two modules to HAPI: `hapi-fhir-jpaserver-empi` and  `hapi-fhir-server-empi`
* New `EmpiLink` entity for an easily accessible representation of links.
* New FHIR operations for operating on links. 
* Rule-matching system for matching patients to other patients based on configurable attributes. 
* Concept of internal and external Enterprise Identifier systems (EID). 
* New subscriptions to handle EMPI processing post-create and post-update of a relevant resource. 
* Modification of ResourceModifiedMessage to add the ability to do distributed transaction-log tracking.
* Lots of docs (but we still need many more docs)
*  120 tests or so :tada:

### CODE REVIEW

File: hapi-fhir-jpaserver-empi/src/main/java/ca/uhn/fhir/jpa/empi/svc/EmpiPersonFindingSvc.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+	 */
+	private List<MatchedPersonCandidate> attemptToFindPersonCandidateFromEmpiLinkTable(IAnyResource theBaseResource) {
+		List<MatchedPersonCandidate> retval = new ArrayList<>();
+
+		Long targetPid = myIdHelperService.getPidOrNull(theBaseResource);
+		if (targetPid != null) {
+			Optional<EmpiLink> oLink = myEmpiLinkDaoSvc.getMatchedLinkForTargetPid(targetPid);
+			if (oLink.isPresent()) {
+				ResourcePersistentId personPid = new ResourcePersistentId(oLink.get().getPersonPid());
+				ourLog.debug("Resource previously linked.  Using existing link.");
```
Comment (jamesagnew) [inline hapi-fhir-jpaserver-empi/src/main/java/ca/uhn/fhir/jpa/empi/svc/EmpiPersonFindingSvc.java:122]: Ooh two spaces after the period. So uhhhhhhhhhhhhhhhhhhhhhhhhhhh the AP style guide doesn't recommend two spaces after a period so clearly like we can't have that geez

File: hapi-fhir-jpaserver-empi/src/test/java/ca/uhn/fhir/jpa/empi/matcher/BasePersonMatcher.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public abstract class BasePersonMatcher extends TypeSafeMatcher<IAnyResource> {
```
Comment (jamesagnew) [inline hapi-fhir-jpaserver-empi/src/test/java/ca/uhn/fhir/jpa/empi/matcher/BasePersonMatcher.java:19]: These typesafematchers are really cool!

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/context/FhirContext.java [hunk 3]
```diff
@@ -502,7 +547,7 @@ public RuntimeResourceDefinition getResourceDefinitionById(final String theId) {
 	 * Returns an unmodifiable set containing all resource names known to this
 	 * context
 	 */
-	public Set<String> getResourceNames() {
+	public Set<String> getResourceTypes() {
```
[the remark this answers is not in the document]
Comment (tadgh) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/context/FhirContext.java:552]: Resolved in 82b3756ffd9db6b48dac4acfd1caa841536e69dd

File: hapi-fhir-jpaserver-empi/src/main/java/ca/uhn/fhir/jpa/empi/svc/EmpiPersonFindingSvc.java (hunk 1 above)
Comment (fil512) [reply hapi-fhir-jpaserver-empi/src/main/java/ca/uhn/fhir/jpa/empi/svc/EmpiPersonFindingSvc.java:122]: I learned to type in 1979 on a manual typewriter and had two spaces DRILLED into my soul.  I have tried.  I cannot change this.  It is too deeply engrained.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/config/BaseConfig.java [hunk 4]
```diff
@@ -102,7 +102,8 @@
 	@ComponentScan.Filter(type = FilterType.REGEX, pattern = ".*\\.test\\..*"),
 	@ComponentScan.Filter(type = FilterType.REGEX, pattern = ".*Test.*"),
 	@ComponentScan.Filter(type = FilterType.REGEX, pattern = "ca.uhn.fhir.jpa.subscription.*"),
-	@ComponentScan.Filter(type = FilterType.REGEX, pattern = "ca.uhn.fhir.jpa.searchparam.*")
+	@ComponentScan.Filter(type = FilterType.REGEX, pattern = "ca.uhn.fhir.jpa.searchparam.*"),
+	@ComponentScan.Filter(type = FilterType.REGEX, pattern = "ca.uhn.fhir.jpa.empi.*")
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/config/BaseConfig.java:107]: This is an exclude.  Empi does not `@ComponentScan`.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/EmpiLinkDaoSvc.java [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+	public Optional<EmpiLink> getLinkByPersonPidAndTargetPid(Long thePersonPid, Long theTargetPid) {
+
+		if (theTargetPid == null || thePersonPid == null) {
+			return Optional.empty();
+		}
+		EmpiLink link = new EmpiLink();
+		link.setTargetPid(theTargetPid);
+		link.setPersonPid(thePersonPid);
+		Example<EmpiLink> example = Example.of(link);
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/EmpiLinkDaoSvc.java:91]: It becomes particularly useful in the query empi link where there are 4 parameters that are all nullable and you want to be able to query by all 4 of them.  Beats 16 `@Query` methods.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/entity/EmpiLink.java [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+	private EmpiLinkSourceEnum myLinkSource;
+
+	@Temporal(TemporalType.TIMESTAMP)
+	@Column(name = "CREATED", nullable = false)
+	@OptimisticLock(excluded = true)
+	private Date myCreated;
+
+	@Temporal(TemporalType.TIMESTAMP)
+	@Column(name = "UPDATED", nullable = false)
+	@OptimisticLock(excluded = true)
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/entity/EmpiLink.java:93]: I just copy/pasted those from elsewhere.  Gary asked me what it did and I had no idea.  Gary wondered if there was a way to automatically maintain these columns.  If that's what `@Version` does then we should do this properly

File: hapi-fhir-jpaserver-empi/src/main/java/ca/uhn/fhir/jpa/empi/svc/EmpiPersonFindingSvc.java (hunk 1 above)
Comment (tadgh) [reply hapi-fhir-jpaserver-empi/src/main/java/ca/uhn/fhir/jpa/empi/svc/EmpiPersonFindingSvc.java:122]: resolved in 78babeb104e176af843a903bf1d5c98cdb472cf7

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/util/StringNormalizer.java [hunk 7]
```diff
@@ -1,8 +1,8 @@
-package ca.uhn.fhir.jpa.model.util;
+package ca.uhn.fhir.util;
 
 /*-
  * #%L
- * HAPI FHIR Model
+ * HAPI FHIR - Core Library
```
[the remark this answers is not in the document]
Comment (tadgh) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/util/StringNormalizer.java:5]: Resolved in ec2b169b7e75644bd1e85cb314bb7dd87cfc786e

File: hapi-fhir-jpaserver-empi/src/main/java/ca/uhn/fhir/jpa/empi/svc/EmpiPersonFindingSvc.java (hunk 1 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-empi/src/main/java/ca/uhn/fhir/jpa/empi/svc/EmpiPersonFindingSvc.java]: hahahaha

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/entity/EmpiLink.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/entity/EmpiLink.java]: We took another look and decided to just delete these annotations

[the remark this answers is not in the document]
Comment (tadgh) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/entity/EmpiLink.java]: Removed in 8ce96f755697ac4b799c34962ac7b3495be10805

[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/entity/EmpiLink.java]: @Version creates an optimistic lock column, which is useful if you think that both:

* there is a chance for two threads to update a row with different values at the same time
* this happening would be a bad thing

...or....

* you want to make sure that two threads are never touching the same row at the same time

For system processing like this I'd say it's less common for that to be the case. Not unheard of at all though. We have an @Version on `BulkExportJobEntity` for the second reason for example

File: hapi-fhir-jpaserver-empi/src/main/java/ca/uhn/fhir/jpa/empi/svc/EmpiCandidateSearchSvc.java [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+	@Nonnull
+	private String buildResourceQueryString(String theResourceType, List<String> theFilterCriteria, EmpiResourceSearchParamJson resourceSearchParam, List<String> theValuesFromResourceForSearchParam) {
+		List<String> criteria = new ArrayList<>(theFilterCriteria);
+		criteria.add(buildResourceMatchQuery(resourceSearchParam.getSearchParam(), theValuesFromResourceForSearchParam));
+
+		return theResourceType + "?" +  String.join("&", criteria);
+	}
+
+	private String buildResourceMatchQuery(String theSearchParamName, List<String> theResourceValues) {
+		return theSearchParamName + "=" + String.join(",", theResourceValues);
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-jpaserver-empi/src/main/java/ca/uhn/fhir/jpa/empi/svc/EmpiCandidateSearchSvc.java:157]: Good catch!  Fixed

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/config/BaseConfig.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/config/BaseConfig.java:107]: Sorry, missed that fact. Resolving this.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/EmpiLinkDaoSvc.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/EmpiLinkDaoSvc.java:91]: Fair enough. I guess I'll restate this comment to just be:

Make sure you have appropriate indexes for any columns being searched on :)

[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/EmpiLinkDaoSvc.java:91]: Following the rule of thumb "If you are reading more than 5% of the records from a table, then a full table scan will be faster than an indexed lookup." We'll index the two id columns but leave the two enum columns unindexed.

[the remark this answers is not in the document]
Comment (tadgh) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/EmpiLinkDaoSvc.java:91]: Resolved in 6a6111ab6a178c79306a0224e5d648660bd7ac09

**Pull request merged**